### PR TITLE
Reset password logic + reset modal

### PR DESCRIPTION
**Pull request by remuskwan**
**Note**: have yet to test update of user using prisma, as CRUD of `User` has yet to be completed
- Created a temporary forgot password page to demo forgot password link + forgot password modal at {domain}/forgot-password (actual link will be in login page)
forgot-password-page
- A password reset link will be sent to email address entered if it exists in the db.
- Public reset token is generated and a hashed version is stored as `passwordResetToken` in the `User` (updated `schema.prisma` for my use case)
```prisma
model User {
  ...
  passwordResetToken String? @unique
  passwordResetAt    DateTime?
  ...
  @@index([passwordResetToken])
  ...
}
```
- Reset token appended to URL of reset link (open to suggestions for improvements on this design)
ethereal-email
- Used fake SMTP service for dev purposes.
- Reset link redirects user to reset password page at {domain}/reset-password/{reset-token}, where they can enter and update their new password
- Added basic input validation, awaiting Formik integration.
reset-password

### CODE REVIEW

**Review by MorningLit — APPROVED** [review]
lgtm for frontend part, but currently facing the prisma db error so didn't had time to test out that part, i just assume it works
just need some small email validation for forget-password and nice modal component too 😍

File: pages/reset-password/[token].tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      console.error(response.status);
+    }
+    const data = await response.json();
+    //TODO: replace with redirect to login page
+    console.log('POST ', data);
+  };
+  const canSubmit = newPassword && confirmPassword;
+  return (
+    <Layout title='Reset Password | Next.js + TypeScript Example'>
+      {/* TODO: replace with Formik */}
```
Comment (MorningLit) [inline pages/reset-password/[token].tsx:41]: replace with react-hook-form

**Comment by weixin-tan** [comment]
Brain fart + sausage finger on the PR closing. Please ignore. Thanks!

File: lib/server/user.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { Prisma, User } from '@prisma/client';
+import prisma from '../prisma';
+//https://codevoweb.com/crud-api-node-prisma-postgresql-reset-password/
+export const findUser = async (where: Partial<Prisma.UserWhereInput>, select?: Prisma.UserSelect) => {
```
Comment (kevinchua6) [inline lib/server/user.ts:4]: Nice use of types! :)

File: pages/api/auth/forgotpassword/index.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        message,
+      });
+    }
+    const resetToken = randomBytes(32).toString('hex');
+    const passwordResetToken = createHash('sha256').update(resetToken).digest('hex');
+    await updateUser(
+      { id: user.id },
+      {
+        passwordResetToken,
+        passwordResetAt: new Date(Date.now() + 10 * 60 * 1000),
```
Comment (kevinchua6) [inline pages/api/auth/forgotpassword/index.ts:24]: If it's not too much trouble can you convert this to dayjs? Thanks!

Comment (kevinchua6) [reply pages/api/auth/forgotpassword/index.ts:24]: And maybe change the name to `passwordResetExpiryDate` or something, I was a bit confused when you checked `gt` later down in your code

File: pages/forgot-password/index.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  const { isOpen, onOpen, onClose } = useDisclosure();
+
+  const handleEmailChanged = (evt: FormEvent<HTMLInputElement>) => setEmail(evt.currentTarget.value);
+  const handleModalClosed = () => {
+    setMessage(false);
+    onClose();
+  };
+  const handleSubmit = async (evt: SyntheticEvent) => {
+    evt.preventDefault();
+    setIsLoading(true);
```
Comment (kevinchua6) [inline pages/forgot-password/index.tsx:19]: We have to remember to change the loading state here when react-query is added (in @kxrt's PR)

File: pages/forgot-password/index.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    try {
+      const response = await fetch('/api/auth/forgotpassword', {
```
Comment (kevinchua6) [inline pages/forgot-password/index.tsx:21]: And use axios here as well

**Review by kevinchua6 — CHANGES_REQUESTED** [review]
Good job on this PR! :) Just a few lines to change

File: pages/forgot-password/index.tsx (hunk 4 above)
Comment (kevinchua6) [reply pages/forgot-password/index.tsx:19]: Can you help to create an issue for this? Just a note for us to change all loading states to the one given by react-query

**Review by kevinchua6 — APPROVED** [review]
LGTM

**Pull request merged**